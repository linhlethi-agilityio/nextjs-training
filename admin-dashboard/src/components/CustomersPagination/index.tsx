// Api
import { getCustomers } from '@/api';

// Constants
import { DEFAULT_LIMIT } from '@/constants';

// Components
import { Pagination } from '@/components';

interface ProductPaginationProps {
  limit?: number;
  query?: string;
}

const CustomersPagination = async ({
  limit = DEFAULT_LIMIT,
  query = '',
}: ProductPaginationProps) => {
  const { data: customers = [] } = await getCustomers({ query });

  const totalPage = Math.ceil(customers?.length / limit);

  return customers.length !== 0 && <Pagination totalPage={totalPage} />;
};

export default CustomersPagination;
