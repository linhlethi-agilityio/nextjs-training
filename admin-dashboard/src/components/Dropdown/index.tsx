'use client';

import { ChangeEvent, memo } from 'react';
import { Select, SelectProps } from '@chakra-ui/react';
import isEqual from 'react-fast-compare';

export interface Option {
  value: string;
  label: string;
}

interface DropdownProps extends SelectProps {
  options: Option[];
  color?: string;
  placeholder?: string;
  onChangeValue: (value: string) => void;
}

const DropdownUI = ({
  options,
  placeholder = '',
  color = 'textDefault',
  width = 108,
  onChangeValue,
  ...rest
}: DropdownProps) => {
  /**
   * Function handle select dropdown
   */
  const handleSelect = (e: ChangeEvent<HTMLSelectElement>) => {
    const newValue = e.target.value;

    onChangeValue(newValue);
  };

  return (
    <Select
      placeholder={placeholder}
      color={color}
      aria-label="Select"
      onChange={handleSelect}
      lineHeight={2}
      w={width}
      {...rest}
    >
      {options.map((option) => {
        const { value = '', label = '' } = option;

        return (
          <option key={value} value={value}>
            {label}
          </option>
        );
      })}
    </Select>
  );
};

export const Dropdown = memo(DropdownUI, isEqual);
