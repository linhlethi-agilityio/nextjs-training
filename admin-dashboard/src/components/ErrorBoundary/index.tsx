'use client';

import { useEffect } from 'react';
import { Button, Flex, Heading } from '@chakra-ui/react';

interface ErrorProps {
  error: Error & { digest?: string };
  reset: () => void;
}

const ErrorBoundary = ({ error, reset }: ErrorProps) => {
  // Attempt to recover by trying to re-render the segment
  const handleReset = () => {
    reset();
  };

  useEffect(() => {
    // Log the error to an error reporting service
    console.error(error);
  }, [error]);

  return (
    <Flex flexDir="column" alignItems="center" gap="50px" mt={300}>
      <Heading>Something went wrong!</Heading>
      <Button onClick={handleReset}>Try again</Button>
    </Flex>
  );
};

export default ErrorBoundary;
