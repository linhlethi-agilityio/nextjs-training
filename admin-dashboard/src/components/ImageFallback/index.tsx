'use client';

import Image, { ImageProps } from 'next/image';
import { memo, useState } from 'react';

interface ImageFallbackProps extends ImageProps {
  src: string;
  fallbackSrc: string;
  alt: string;
}

const ImageFallback = ({
  src,
  fallbackSrc,
  alt,
  ...rest
}: ImageFallbackProps) => {
  const [imgSrc, setImgSrc] = useState(src);

  return (
    <Image
      src={imgSrc}
      blurDataURL={fallbackSrc}
      placeholder="blur"
      onLoadingComplete={(result) => {
        if (result.naturalWidth === 0) {
          // Broken image
          setImgSrc(fallbackSrc);
        }
      }}
      onError={() => {
        setImgSrc(fallbackSrc);
      }}
      alt={alt}
      {...rest}
    />
  );
};

export default memo(ImageFallback);
