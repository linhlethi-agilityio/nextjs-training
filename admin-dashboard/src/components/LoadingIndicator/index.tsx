import { Box, Spinner as SpinnerChakraUI } from '@chakra-ui/react';

const LoadingIndicator = (): JSX.Element => {
  return (
    <Box
      h="100vh"
      display="flex"
      alignItems="center"
      textAlign="center"
      justifyContent="center"
    >
      <SpinnerChakraUI
        size="lg"
        thickness="4px"
        speed="0.65s"
        color="brand.500"
      />
    </Box>
  );
};

export default LoadingIndicator;
