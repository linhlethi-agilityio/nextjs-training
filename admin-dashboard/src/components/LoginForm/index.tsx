'use client';

import { useMemo, KeyboardEvent, useState, memo, useTransition } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { useRouter } from 'next/navigation';
import {
  Box,
  Button,
  FormControl,
  FormHelperText,
  Input,
  InputGroup,
  InputLeftElement,
  InputRightElement,
  useToast,
} from '@chakra-ui/react';

// Constants
import { ERROR_MESSAGES, ROUTES, SUCCESS_MESSAGES } from '@/constants';

// Utils
import {
  clearErrorOnChange,
  isEnableSubmitButton,
  isValidEmail,
} from '@/utils';

// Icons
import { EmailIcon, HideEyeIcon, PasswordIcon, ShowEyeIcon } from '@/icons';

interface LoginFormData {
  email: string;
  password: string;
}

const REQUIRED_FIELDS = ['email', 'password'];

interface LoginForm {
  onSubmit: (data: LoginFormData) => Promise<void | string>;
}

const LoginForm = ({ onSubmit }: LoginForm) => {
  const [showPassword, setShowPassword] = useState<boolean>(false);

  const toast = useToast();
  const [isPending, startTransition] = useTransition();
  const router = useRouter();

  const {
    control,
    formState: { dirtyFields, errors },
    clearErrors,
    handleSubmit,
  } = useForm<LoginFormData>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: {
      email: '',
      password: '',
    },
  });

  // Checking to disable/enable submit button
  const dirtyItems = Object.keys(dirtyFields);

  const enableSubmit: boolean = useMemo(
    () => isEnableSubmitButton(REQUIRED_FIELDS, dirtyItems, errors),
    [dirtyItems, errors],
  );
  const isDisableSubmit = !enableSubmit;

  /**
   * Func handle login
   */
  const handleLogin = async (formData: LoginFormData) => {
    startTransition(async () => {
      const response = await onSubmit(formData);
      console.log('');

      if (typeof response === 'string') {
        toast({
          title: response,
          status: 'error',
        });
      } else {
        toast({
          title: SUCCESS_MESSAGES.LOGIN_SUCCESS,
          status: 'success',
        });
        router.push(ROUTES.PRODUCT);
      }
    });
  };

  const handleClickEyePassword = () => setShowPassword(!showPassword);

  const handleOnKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'Enter') {
      handleSubmit(handleLogin)();
    }
  };

  return (
    <FormControl mt={50} onKeyDown={handleOnKeyDown} maxW={500}>
      <Controller
        name="email"
        control={control}
        rules={{
          required: ERROR_MESSAGES.FIELD_REQUIRED,
          validate: (email) =>
            isValidEmail(email) || ERROR_MESSAGES.INVALID_EMAIL,
        }}
        render={({
          field: { name, onChange, ...rest },
          fieldState: { error },
        }) => (
          <Box
            marginBottom={error?.message ? 0 : 26}
            _placeholder={{ color: 'textPlaceholder' }}
            ml={3}
          >
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <EmailIcon color="textPlaceholder" />
              </InputLeftElement>
              <Input
                id="email"
                placeholder="Your email"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
            </InputGroup>

            {error?.message && (
              <FormHelperText
                color="red.400"
                fontSize="xs"
                mb={1}
                textAlign="left"
              >
                {error.message}
              </FormHelperText>
            )}
          </Box>
        )}
      />

      {/* Password */}
      <Controller
        name="password"
        control={control}
        rules={{
          required: ERROR_MESSAGES.FIELD_REQUIRED,
        }}
        render={({
          field: { name, onChange, ...rest },
          fieldState: { error },
        }) => (
          <Box
            marginBottom={error?.message ? 0 : 26}
            ml={3}
            _placeholder={{ color: 'textPlaceholder' }}
          >
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <PasswordIcon />
              </InputLeftElement>
              <InputRightElement>
                <Button variant="icon" onClick={handleClickEyePassword}>
                  {showPassword ? <ShowEyeIcon /> : <HideEyeIcon />}
                </Button>
              </InputRightElement>
              <Input
                w="full"
                id="password"
                placeholder="Password"
                type={showPassword ? 'text' : 'password'}
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
            </InputGroup>

            {error?.message && (
              <FormHelperText
                color="red.400"
                fontSize="xs"
                mb={1}
                textAlign="left"
              >
                {error.message}
              </FormHelperText>
            )}
          </Box>
        )}
      />

      <Button
        isDisabled={isDisableSubmit}
        isLoading={isPending}
        colorScheme="brand"
        ml={3}
        aria-label="log-in"
        w="-webkit-fill-available"
        mt={2}
        onClick={handleSubmit(handleLogin)}
      >
        Log In
      </Button>
    </FormControl>
  );
};

export default memo(LoginForm);
