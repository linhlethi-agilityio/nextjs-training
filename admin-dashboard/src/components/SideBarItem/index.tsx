import { memo } from 'react';
import { Flex, HStack, IconProps, Text } from '@chakra-ui/react';
import Link from 'next/link';

export interface SideBarItemProps {
  isFocused?: boolean;
  icon: React.FC<IconProps>;
  label: string;
  path: string;
}

const SideBarItem = ({
  label = '',
  isFocused = false,
  icon,
  path,
}: SideBarItemProps) => {
  const Icon = icon;

  return (
    <Link href={path} style={{ width: '-webkit-fill-available' }}>
      <HStack
        data-testid={`sidebar-item-${label}`}
        spacing={isFocused ? 2 : 3}
        py={3.5}
        pl={15}
        transition="1s"
        color={isFocused ? 'textPrimary' : 'textDefault'}
        cursor="pointer"
        {...(isFocused && {
          border: '1px solid',
          borderRadius: '2xl',
          borderColor: 'borderPrimary',
        })}
      >
        <Flex
          {...(isFocused && {
            p: 1,
            bgColor: 'backgroundSiBarItem',
            borderRadius: 'base',
          })}
        >
          <Icon w={4} h={4} />
        </Flex>
        <Text color={isFocused ? 'textPrimary' : 'textDefault'} size="xs">
          {label}
        </Text>
      </HStack>
    </Link>
  );
};

export default memo(SideBarItem);
