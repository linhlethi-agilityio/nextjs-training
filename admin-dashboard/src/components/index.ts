export { default as SearchInput } from './SearchInput';
export { default as SideBarItem } from './SideBarItem';
export { default as SideBar } from './SideBar';
export { default as Pagination } from './Pagination';
export { default as ErrorBoundary } from './ErrorBoundary';
export { default as LoadingIndicator } from './LoadingIndicator';
export { default as ProductActions } from './ProductActions';
export { default as ProductPagination } from './ProductPagination';
export { default as LoginForm } from './LoginForm';
export { default as RegisterForm } from './RegisterForm';
export { default as OrderDetail } from './OrderDetail';
export { default as TableCustomers } from './TableCustomers';
export * from './Dropdown';

// Modals
export * from './Modal/CustomModal';
export { default as ConfirmModal } from './Modal/ConfirmModal';
export { default as OrderModal } from './Modal/OrderModal';

// Skeleton
export * from './Skeleton';

// Table
export { default as TableCell } from './TableCell';
export { default as TableOrder } from './TableOrder';
export * from './TableRow';
export * from './Table';
