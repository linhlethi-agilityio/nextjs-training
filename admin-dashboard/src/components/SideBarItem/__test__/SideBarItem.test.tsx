import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

// Icons
import { FileTextFillIcon } from '@/icons';

// Components
import { SideBarItem } from '@/components';

const mockProps = {
  icon: FileTextFillIcon,
  path: '/',
  label: 'Product',
};

describe('SideBarItem component', () => {
  it('should match snapshot for SideBarItem', () => {
    const { container } = render(<SideBarItem {...mockProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for SideBarItem', () => {
    const { container } = render(<SideBarItem {...mockProps} isFocused />);

    expect(container).toMatchSnapshot();
  });
});
