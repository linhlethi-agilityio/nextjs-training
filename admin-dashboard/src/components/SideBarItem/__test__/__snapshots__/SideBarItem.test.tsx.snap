// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBarItem component should match snapshot for SideBarItem 1`] = `
<div>
  <a
    href="/"
  >
    <div
      class="chakra-stack css-ile8xj"
      data-testid="sidebar-item-Product"
    >
      <div
        class="css-k008qs"
      >
        <svg
          class="chakra-icon css-z2vag5"
          focusable="false"
          viewBox="0 0 10 12"
        >
          <path
            d="M8.33057 0H2.33057C1.93274 0 1.55121 0.158035 1.26991 0.43934C0.988602 0.720644 0.830566 1.10218 0.830566 1.5V10.5C0.830566 10.8978 0.988602 11.2794 1.26991 11.5607C1.55121 11.842 1.93274 12 2.33057 12H8.33057C8.72839 12 9.10992 11.842 9.39123 11.5607C9.67253 11.2794 9.83057 10.8978 9.83057 10.5V1.5C9.83057 1.10218 9.67253 0.720644 9.39123 0.43934C9.10992 0.158035 8.72839 0 8.33057 0ZM3.08057 3H7.58057C7.68002 3 7.77541 3.03951 7.84573 3.10984C7.91606 3.18016 7.95557 3.27554 7.95557 3.375C7.95557 3.47446 7.91606 3.56984 7.84573 3.64016C7.77541 3.71049 7.68002 3.75 7.58057 3.75H3.08057C2.98111 3.75 2.88573 3.71049 2.8154 3.64016C2.74508 3.56984 2.70557 3.47446 2.70557 3.375C2.70557 3.27554 2.74508 3.18016 2.8154 3.10984C2.88573 3.03951 2.98111 3 3.08057 3ZM2.70557 4.875C2.70557 4.77554 2.74508 4.68016 2.8154 4.60984C2.88573 4.53951 2.98111 4.5 3.08057 4.5H7.58057C7.68002 4.5 7.77541 4.53951 7.84573 4.60984C7.91606 4.68016 7.95557 4.77554 7.95557 4.875C7.95557 4.97446 7.91606 5.06984 7.84573 5.14016C7.77541 5.21049 7.68002 5.25 7.58057 5.25H3.08057C2.98111 5.25 2.88573 5.21049 2.8154 5.14016C2.74508 5.06984 2.70557 4.97446 2.70557 4.875ZM3.08057 6H7.58057C7.68002 6 7.77541 6.03951 7.84573 6.10983C7.91606 6.18016 7.95557 6.27554 7.95557 6.375C7.95557 6.47446 7.91606 6.56984 7.84573 6.64017C7.77541 6.71049 7.68002 6.75 7.58057 6.75H3.08057C2.98111 6.75 2.88573 6.71049 2.8154 6.64017C2.74508 6.56984 2.70557 6.47446 2.70557 6.375C2.70557 6.27554 2.74508 6.18016 2.8154 6.10983C2.88573 6.03951 2.98111 6 3.08057 6ZM3.08057 7.5H5.33057C5.43002 7.5 5.52541 7.53951 5.59573 7.60983C5.66606 7.68016 5.70557 7.77554 5.70557 7.875C5.70557 7.97446 5.66606 8.06984 5.59573 8.14017C5.52541 8.21049 5.43002 8.25 5.33057 8.25H3.08057C2.98111 8.25 2.88573 8.21049 2.8154 8.14017C2.74508 8.06984 2.70557 7.97446 2.70557 7.875C2.70557 7.77554 2.74508 7.68016 2.8154 7.60983C2.88573 7.53951 2.98111 7.5 3.08057 7.5Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <p
        class="chakra-text css-vpapf9"
      >
        Product
      </p>
    </div>
  </a>
</div>
`;

exports[`SideBarItem component should match snapshot for SideBarItem 2`] = `
<div>
  <a
    href="/"
  >
    <div
      class="chakra-stack css-9aktsg"
      data-testid="sidebar-item-Product"
    >
      <div
        class="css-8z21bf"
      >
        <svg
          class="chakra-icon css-z2vag5"
          focusable="false"
          viewBox="0 0 10 12"
        >
          <path
            d="M8.33057 0H2.33057C1.93274 0 1.55121 0.158035 1.26991 0.43934C0.988602 0.720644 0.830566 1.10218 0.830566 1.5V10.5C0.830566 10.8978 0.988602 11.2794 1.26991 11.5607C1.55121 11.842 1.93274 12 2.33057 12H8.33057C8.72839 12 9.10992 11.842 9.39123 11.5607C9.67253 11.2794 9.83057 10.8978 9.83057 10.5V1.5C9.83057 1.10218 9.67253 0.720644 9.39123 0.43934C9.10992 0.158035 8.72839 0 8.33057 0ZM3.08057 3H7.58057C7.68002 3 7.77541 3.03951 7.84573 3.10984C7.91606 3.18016 7.95557 3.27554 7.95557 3.375C7.95557 3.47446 7.91606 3.56984 7.84573 3.64016C7.77541 3.71049 7.68002 3.75 7.58057 3.75H3.08057C2.98111 3.75 2.88573 3.71049 2.8154 3.64016C2.74508 3.56984 2.70557 3.47446 2.70557 3.375C2.70557 3.27554 2.74508 3.18016 2.8154 3.10984C2.88573 3.03951 2.98111 3 3.08057 3ZM2.70557 4.875C2.70557 4.77554 2.74508 4.68016 2.8154 4.60984C2.88573 4.53951 2.98111 4.5 3.08057 4.5H7.58057C7.68002 4.5 7.77541 4.53951 7.84573 4.60984C7.91606 4.68016 7.95557 4.77554 7.95557 4.875C7.95557 4.97446 7.91606 5.06984 7.84573 5.14016C7.77541 5.21049 7.68002 5.25 7.58057 5.25H3.08057C2.98111 5.25 2.88573 5.21049 2.8154 5.14016C2.74508 5.06984 2.70557 4.97446 2.70557 4.875ZM3.08057 6H7.58057C7.68002 6 7.77541 6.03951 7.84573 6.10983C7.91606 6.18016 7.95557 6.27554 7.95557 6.375C7.95557 6.47446 7.91606 6.56984 7.84573 6.64017C7.77541 6.71049 7.68002 6.75 7.58057 6.75H3.08057C2.98111 6.75 2.88573 6.71049 2.8154 6.64017C2.74508 6.56984 2.70557 6.47446 2.70557 6.375C2.70557 6.27554 2.74508 6.18016 2.8154 6.10983C2.88573 6.03951 2.98111 6 3.08057 6ZM3.08057 7.5H5.33057C5.43002 7.5 5.52541 7.53951 5.59573 7.60983C5.66606 7.68016 5.70557 7.77554 5.70557 7.875C5.70557 7.97446 5.66606 8.06984 5.59573 8.14017C5.52541 8.21049 5.43002 8.25 5.33057 8.25H3.08057C2.98111 8.25 2.88573 8.21049 2.8154 8.14017C2.74508 8.06984 2.70557 7.97446 2.70557 7.875C2.70557 7.77554 2.74508 7.68016 2.8154 7.60983C2.88573 7.53951 2.98111 7.5 3.08057 7.5Z"
            fill="currentColor"
          />
        </svg>
      </div>
      <p
        class="chakra-text css-zz3tyx"
      >
        Product
      </p>
    </div>
  </a>
</div>
`;
